text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+equivalent to `pub impl Foo for Bar { .. }`. This is right for two reasons:
+
++ Changing the default visibility to module private would constitute a major
+  breaking change which would be too large to countenance even with an edition.
+
++ `pub` as a default for implementations is a *good* default, since it is what
+  you want most of the time.
+
+Having `pub` be permitted on `impl` also means that you can use the `vis`
+macro fragment specifier.
+
+To promote uniformity in the ecosystem, directly writing `pub impl ..` will
+be linted against with a lint named `redundant_vis` (or a naming in that spirit).
+A sample warning is:
+
+```rust
+warning: `pub` is redundant on `impl`
+ --> src/main.rs:<line>:<column>
+  |
+L |     pub impl Foo for Bar { .. }
+  |     ^^^ help: remove `pub`
+  |
+  = note: #[warn(redundant_vis)] on by default
+```
+
+## `priv` visibility
+
+To be able to make implementations module private, this RFC introduces a new
+visibility modifier called `priv` (which is currently a reserved keyword).
+This new visibility modifier is made the default for all contexts where
+visibility modifiers are permitted *but* in the context of implementations
+(where `pub` is the default). Directly writing:
+
+```rust
+struct Foo {
+    priv x: Bar,
+}
+```
+
+will be linted against with `redundant_vis` just like `pub impl ..`
+will be linted against.
+
+## `#[derive(crate Trait)]`
+
+To make it simpler for you to provide implementation of common traits
+which can be derived, you are permitted to place a visibility modifier,
+such as `crate`, before `Trait` in the `#[derive(..)]` attribute on a
+type definition. An example:
+
+```rust
+#[derive(Default, crate Copy, crate Clone)]
+struct Foo {
+    ...
+}
+```
+
+In the case of derivable standard library traits, the compiler will insert the
+visibility modifier before `impl` in the implementations generated.
+Custom derive macros are not obligated to respect this modifier.
+However, it is nonetheless recommended that such macros,
+which are available for public use, handle these modifiers,
+or emit an error notifying the user that modifiers are not supported.
+
+## In relation to trait objects
+
+One place where crate B is allowed to use a hidden implementation of crate A
+is when it comes to trait objects. If A provides access to a trait object
+`Box<dyn Property>` (which implies that `Property` is object safe) by returning
+a value of that type, the compiler has no means in the general case to perform
+static analysis on that object to see where the implementation came from.
+This is fine. In this case, it is crate A that expressly gave access to an
+implementation of `Property` for `Thing`. Crate A is also under no obligation
+under the semver rules to provide an erased `Thing` value. As long as the value
+returned behaves observably in an equivalent manner, for some semantic
+definition of "equivalent", no breakage has occurred.
```

Comment:
So essentially, `return Box::new(Thing): Box<dyn Property>` is permitted with no compiler warnings? I'm confused why the compiler *cannot* lint against this -- it knows that it is creating a vtable for the impl of Property for Thing; as such that place can lint against this (though in many cases only after/during monomorphization).

However, I'm not opposed to this (just feels like the wording could be changed).